File: scripts/quota-cache/quota_cache_client.py
#!/usr/bin/env python3
"""Minimal consumer client for the Quotio quota-cache service.

Shared by `cpa_quota_recovery.py` and `cpa_newapi_balance.py` so both scripts —
and Quotio itself — read one collected value per account/resource instead of
each polling CLIProxyAPI (and, transitively, the real provider) on its own
schedule. Python 3 standard library only, matching every other script here.
"""

from __future__ import annotations

import json
import math
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict


class QuotaCacheError(Exception):
    """A cache failure with an optional fixed, non-sensitive account category."""

    def __init__(
        self,
        message: str,
        *,
        http_status: int | None = None,
        failure_kind: str | None = None,
        failure_status_code: int | None = None,
    ) -> None:
        super().__init__(message)
        self.http_status = http_status
        self.failure_kind = failure_kind
        self.failure_status_code = failure_status_code

    @property
    def is_auth_invalid(self) -> bool:
        return self.failure_kind == "auth_invalid" and self.failure_status_code in (401, 403)


def _safe_failure(body: bytes) -> tuple[str | None, int | None]:
    try:
        payload = json.loads(body.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        return None, None
    failure = payload.get("failure") if isinstance(payload, dict) else None
    if not isinstance(failure, dict) or failure.get("kind") != "auth_invalid":
        return None, None
    status = failure.get("status_code")
    if isinstance(status, bool) or status not in (401, 403):
        return None, None
    return "auth_invalid", status


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, *args: Any, **kwargs: Any) -> None:  # noqa: D401
        return None


def fetch(
    cache_base_url: str,
    management_key: str,
    resource: str,
    auth_index: str,
    *,
    require_fresh: bool = False,
    timeout: float = 10.0,
) -> Dict[str, Any]:
    """GETs `{cache_base_url}/{resource}?auth_index=...[&require_fresh=1]` and
    returns the decoded envelope: `{"result": {"status_code", "header", "body"},
    "fetched_at", "stale", "last_attempt", "next_retry_at"}`.

    `require_fresh=True` is for callers that make a decision from the result
    (quota recovery, account weighting) — the service returns HTTP 503 instead
    of stale data when it cannot confirm freshness, which surfaces here as a
    `QuotaCacheError`, never as a fabricated/old result.
    """
    base = cache_base_url.rstrip("/")
    query = {"auth_index": auth_index}
    if require_fresh:
        query["require_fresh"] = "1"
    url = f"{base}/{resource}?{urllib.parse.urlencode(query)}"
    request = urllib.request.Request(
        url, method="GET", headers={"Authorization": f"Bearer {management_key}"}
    )
    opener = urllib.request.build_opener(_NoRedirect())
    try:
        with opener.open(request, timeout=timeout) as response:
            status = response.getcode()
            body = response.read()
    except urllib.error.HTTPError as exc:
        try:
            error_body = exc.read(16_384)
        except Exception:
            error_body = b""
        failure_kind, failure_status_code = _safe_failure(error_body)
        raise QuotaCacheError(
            f"quota-cache returned HTTP {exc.code}",
            http_status=exc.code,
            failure_kind=failure_kind,
            failure_status_code=failure_status_code,
        ) from None
    except urllib.error.URLError as exc:
        raise QuotaCacheError(f"quota-cache unreachable: {exc.reason}") from None
    except OSError as exc:
        raise QuotaCacheError(f"quota-cache request failed: {exc}") from None

    if status != 200:
        failure_kind, failure_status_code = _safe_failure(body)
        raise QuotaCacheError(
            f"quota-cache returned HTTP {status}",
            http_status=status,
            failure_kind=failure_kind,
            failure_status_code=failure_status_code,
        )
    try:
        envelope = json.loads(body.decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as exc:
        raise QuotaCacheError("quota-cache returned invalid JSON") from exc
    if not isinstance(envelope, dict) or not isinstance(envelope.get("result"), dict):
        raise QuotaCacheError("quota-cache returned a malformed envelope")
    result = envelope["result"]
    if not isinstance(result.get("status_code"), int) or isinstance(result.get("status_code"), bool):
        raise QuotaCacheError("quota-cache result is missing a valid status_code")
    if not isinstance(result.get("body"), str):
        raise QuotaCacheError("quota-cache result is missing a body")
    fetched_at = envelope.get("fetched_at")
    if (
        not isinstance(fetched_at, (int, float))
        or isinstance(fetched_at, bool)
        or not math.isfinite(fetched_at)
    ):
        raise QuotaCacheError("quota-cache envelope is missing a valid fetched_at")
    if require_fresh and envelope.get("stale") is not False:
        # Belt-and-suspenders: the service's own contract already guarantees
        # `require_fresh=1` never returns HTTP 200 with `stale: true` (it
        # returns 503 instead — see service.py's `handle`). A caller on this
        # path makes a real decision (quota recovery, account weighting) from
        # the result, so it must never trust a 200 status code alone; a future
        # service bug, a version mismatch, or a misbehaving cache endpoint
        # must still never be silently treated as a confirmed-fresh reading.
        raise QuotaCacheError("quota-cache returned a non-fresh result for a require_fresh request")
    return envelope

File: Packages/QuotioCore/Sources/QuotioInfrastructure/Quota/QuotaCacheClient.swift
import Foundation
import QuotioDomain

/// Reads one whitelisted quota resource through the local quota-cache service
/// (`scripts/quota-cache`) instead of calling a remote CLIProxyAPI's `/api-call`
/// pass-through directly. Injectable via `QuotaHTTPSession` so tests never make a
/// real network request. The service is local-only, read-only, and reuses the same
/// management key already stored for the source — enabling it adds no new secret
/// surface. The client sends only `auth_index` on the query string and the existing
/// management key as a bearer token; it never sends a URL, header, or token for the
/// service to forward, matching the service's own contract that it alone decides
/// which upstream URL/headers a resource maps to.
public struct QuotaCacheClient: Sendable {
    /// The closed resource whitelist mirrored from `scripts/quota-cache/resources.py`'s
    /// `RESOURCES` table. Every call site already passes one of these fixed literals —
    /// this is defense in depth, not a real gate, since nothing derives `resource` from
    /// user input — but it keeps this client refusing to construct a request for
    /// anything the service itself wouldn't recognize either.
    private static let allowedResources: Set<String> = [
        "codex-usage", "codex-reset-credits", "claude-usage", "claude-profile", "grok-usage", "grok-settings",
    ]

    private let session: any QuotaHTTPSession

    public init(session: any QuotaHTTPSession = QuotaCacheClient.makeSession()) {
        self.session = session
    }

    /// A dedicated session, never `URLSession.shared`: this client sends the pool's
    /// management key as a bearer token, so it must never follow a redirect a
    /// compromised/misconfigured cache endpoint could issue to resend that key to an
    /// unrelated host. `AmpNoRedirectDelegate` already exists for exactly this
    /// purpose and denies every redirect unconditionally.
    public nonisolated static func makeSession() -> any QuotaHTTPSession {
        URLSession(
            configuration: ProxyURLSessionFactory.makeConfiguration(timeout: 10),
            delegate: AmpNoRedirectDelegate(),
            delegateQueue: nil
        )
    }

    public func fetch(
        baseURL: String,
        resource: String,
        authIndex: String,
        managementKey: String
    ) async throws -> QuotaCacheResponse {
        guard Self.allowedResources.contains(resource) else {
            throw QuotaCacheError.invalidBaseURL
        }
        guard var components = URLComponents(string: baseURL),
            let scheme = components.scheme?.lowercased(), scheme == "http" || scheme == "https",
            let host = components.host, !host.isEmpty,
            components.user == nil, components.password == nil, components.fragment == nil
        else {
            throw QuotaCacheError.invalidBaseURL
        }
        // `http` is only ever trusted for the documented same-host deployment
        // (`http://127.0.0.1:...`, see scripts/quota-cache/README.md) — anything
        // else must be `https`, so the management key sent as this request's
        // bearer token is never carried in cleartext to a non-loopback host.
        // `RemoteManagementQuotaFetcher` layers a same-origin check for the
        // `https` case on top of this before it ever calls `fetch`.
        guard scheme == "https" || Self.isLoopbackHost(host) else {
            throw QuotaCacheError.invalidBaseURL
        }
        components.path = components.path.hasSuffix("/") ? components.path + resource : components.path + "/" + resource
        components.queryItems = [URLQueryItem(name: "auth_index", value: authIndex)]
        guard let url = components.url else {
            throw QuotaCacheError.invalidBaseURL
        }

        var request = URLRequest(url: url, timeoutInterval: 10)
        request.httpMethod = "GET"
        request.setValue("Bearer \(managementKey)", forHTTPHeaderField: "Authorization")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw QuotaCacheError.connectionError
        }
        guard let http = response as? HTTPURLResponse else {
            throw QuotaCacheError.invalidResponse
        }
        guard http.statusCode == 200 else {
            if let envelope = try? JSONDecoder().decode(QuotaCacheErrorResponse.self, from: data),
               let statusCode = envelope.failure?.authInvalidStatusCode {
                throw QuotaCacheError.authInvalid(statusCode)
            }
            throw QuotaCacheError.httpError(http.statusCode)
        }
        do {
            return try JSONDecoder().decode(QuotaCacheResponse.self, from: data)
        } catch {
            throw QuotaCacheError.invalidResponse
        }
    }

    static func isLoopbackHost(_ host: String) -> Bool {
        let normalized = host.lowercased()
        return normalized == "127.0.0.1" || normalized == "::1" || normalized == "localhost"
    }
}

/// One quota-cache service response envelope. `result` mirrors `ProxyAPICallResult`
/// exactly — the same upstream pass-through shape the direct `apiCall` path already
/// returns — so every existing response-mapping function consumes it unchanged.
/// `fetchedAt` is the real wall-clock time the cached value was last actually
/// obtained from upstream, never the moment this response was read from the cache,
/// so a `Retry-After` header on a stale `result` is still interpreted relative to
/// when it was really received, and `ProviderQuota.lastUpdated` never claims a
/// fresher reading than the cache actually holds.
public struct QuotaCacheResponse: Decodable, Sendable {
    public let result: ProxyAPICallResult
    public let fetchedAt: Double
    public let stale: Bool
    public let lastAttempt: Double
    public let nextRetryAt: Double?
    public let failure: QuotaCacheFailure?
    /// Optional CPA pool routing-weight reading, carried alongside this same
    /// `codex-usage` envelope rather than a separate request — absent entirely (never
    /// decoded as a synthetic zero) on a cache build that doesn't compute weights, or
    /// when this account/pool has none.
    public let routingWeights: QuotaCacheRoutingWeights?

    enum CodingKeys: String, CodingKey {
        case result, stale, failure
        case fetchedAt = "fetched_at"
        case lastAttempt = "last_attempt"
        case nextRetryAt = "next_retry_at"
        case routingWeights = "routing_weights"
    }

    /// Decodes `routingWeights` tolerantly: a malformed `routing_weights` payload
    /// (wrong types, or values `QuotaCacheRoutingWeights` itself rejects) never fails
    /// this whole response's decode — it only means this optional reading comes back
    /// `nil`, exactly like a cache build that omits it entirely. Every other field is
    /// the real quota data and must still fail loudly if it doesn't decode.
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        result = try container.decode(ProxyAPICallResult.self, forKey: .result)
        fetchedAt = try container.decode(Double.self, forKey: .fetchedAt)
        stale = try container.decode(Bool.self, forKey: .stale)
        lastAttempt = try container.decode(Double.self, forKey: .lastAttempt)
        nextRetryAt = try container.decodeIfPresent(Double.self, forKey: .nextRetryAt)
        failure = try container.decodeIfPresent(QuotaCacheFailure.self, forKey: .failure)
        routingWeights = (try? container.decodeIfPresent(QuotaCacheRoutingWeights.self, forKey: .routingWeights)) ?? nil
    }
}

/// Raw wire shape of `QuotaCacheResponse.routingWeights` — `updatedAt` is Unix
/// seconds, matching `fetched_at`/`last_attempt` on the same envelope. Mapped to the
/// Domain `AccountRoutingWeight` by `RemoteManagementQuotaFetcher`, which is the
/// layer that owns converting this raw timestamp into a `Date`.
public struct QuotaCacheRoutingWeights: Decodable, Sendable {
    public let account: Int
    public let channel: Int
    public let updatedAt: Double

    enum CodingKeys: String, CodingKey {
        case account, channel
        case updatedAt = "updated_at"
    }

    /// Rejects a reading the cache should never actually send: a negative weight, or a
    /// non-finite/non-positive `updated_at`. Throwing here (rather than clamping) is
    /// what lets `QuotaCacheResponse`'s tolerant decode turn this into `nil` instead of
    /// silently keeping a nonsensical weight/timestamp.
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let account = try container.decode(Int.self, forKey: .account)
        let channel = try container.decode(Int.self, forKey: .channel)
        let updatedAt = try container.decode(Double.self, forKey: .updatedAt)
        guard account >= 0, channel >= 0, updatedAt.isFinite, updatedAt > 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .updatedAt, in: container, debugDescription: "Invalid routing weights reading"
            )
        }
        self.account = account
        self.channel = channel
        self.updatedAt = updatedAt
    }
}

public struct QuotaCacheFailure: Decodable, Equatable, Sendable {
    public let kind: String
    public let statusCode: Int?

    enum CodingKeys: String, CodingKey {
        case kind
        case statusCode = "status_code"
    }

    var authInvalidStatusCode: Int? {
        guard kind == "auth_invalid", let statusCode, statusCode == 401 || statusCode == 403 else { return nil }
        return statusCode
    }

    var remoteAccountIssue: RemoteQuotaAccountIssue? {
        authInvalidStatusCode == nil ? nil : .invalidOAuth
    }
}

private struct QuotaCacheErrorResponse: Decodable {
    let failure: QuotaCacheFailure?
}

public enum QuotaCacheError: Error, Equatable, Sendable {
    case invalidBaseURL
    case invalidResponse
    case connectionError
    case httpError(Int)
    case authInvalid(Int)
}
